File: src/main/java/im/tox/tox4j/exceptions/ToxException.java
package im.tox.tox4j.exceptions;

import im.tox.tox4j.annotations.NotNull;

public abstract class ToxException extends Exception {

  @NotNull
  public abstract Enum<?> getCode();

  @Override
  @NotNull
  public final String getMessage() {
    return "Error code: " + getCode();
  }

}

File: projects/tox4j/src/main/java/im/tox/tox4j/exceptions/ToxException.scala
package im.tox.tox4j.exceptions

import org.jetbrains.annotations.NotNull

abstract class ToxException[E <: Enum[E]](errorCode: E, message: String) extends Exception(message) {

  @NotNull
  final override def getMessage: String = {
    message match {
      case "" => "Error code: " + errorCode.name
      case _  => message + ", error code: " + errorCode.name
    }
  }

  final def code: E = errorCode

}
